# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import pandas as pd


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather.csv` file. 

#### First import it into a data frame called `austin`.

In [2]:
# Your code here
austin = pd.read_csv('../austin_weather.csv')
austin.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0  2013-12-21         74        60        45            67           49   
1  2013-12-22         56        48        39            43           36   
2  2013-12-23         58        45        32            31           27   
3  2013-12-24         61        46        31            36           28   
4  2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [3]:
# lets see things
austin.describe()
# i can use austin.describe(include=all) to check statistics of al columns

TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

In [4]:
austin.describe(include="all")
# it creates a dataframe with statistics
# top is the mode
# freq is the frequency
# -- division of elements --
# with-str has no unique,top,freq
# numeric: the other rows

Date    TempHighF     TempAvgF     TempLowF DewPointHighF  \
count         1319  1319.000000  1319.000000  1319.000000          1319   
unique        1319          NaN          NaN          NaN            64   
top     2016-10-07          NaN          NaN          NaN            74   
freq             1          NaN          NaN          NaN           112   
mean           NaN    80.862775    70.642911    59.902957           NaN   
std            NaN    14.766523    14.045904    14.190648           NaN   
min            NaN    32.000000    29.000000    19.000000           NaN   
25%            NaN    72.000000    62.000000    49.000000           NaN   
50%            NaN    83.000000    73.000000    63.000000           NaN   
75%            NaN    92.000000    83.000000    73.000000           NaN   
max            NaN   107.000000    93.000000    81.000000           NaN   

       DewPointAvgF DewPointLowF HumidityHighPercent HumidityAvgPercent  \
count          1319         1319                1319               1319   
unique           66           73                  58                 69   
top              71           66                 100                 64   
freq             84           62                 221                 51   
mean            NaN          NaN                 NaN                NaN   
std             NaN          NaN                 NaN                NaN   
min             NaN          NaN                 NaN                NaN   
25%             NaN          NaN                 NaN                NaN   
50%             NaN          NaN                 NaN                NaN   
75%             NaN          NaN                 NaN                NaN   
max             NaN          NaN                 NaN                NaN   

       HumidityLowPercent  ... SeaLevelPressureAvgInches  \
count                1319  ...                      1319   
unique                 82  ...                       101   
top                    52  ...                     29.95   
freq                   46  ...                        49   
mean                  NaN  ...                       NaN   
std                   NaN  ...                       NaN   
min                   NaN  ...                       NaN   
25%                   NaN  ...                       NaN   
50%                   NaN  ...                       NaN   
75%                   NaN  ...                       NaN   
max                   NaN  ...                       NaN   

       SeaLevelPressureLowInches VisibilityHighMiles VisibilityAvgMiles  \
count                       1319                1319               1319   
unique                       105                   5                 10   
top                        29.88                  10                 10   
freq                          48                1302                859   
mean                         NaN                 NaN                NaN   
std                          NaN                 NaN                NaN   
min                          NaN                 NaN                NaN   
25%                          NaN                 NaN                NaN   
50%                          NaN                 NaN                NaN   
75%                          NaN                 NaN                NaN   
max                          NaN                 NaN                NaN   

       VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
count                1319        1319       1319        1319   
unique                 12          22         13          37   
top                    10          13          4          20   
freq                  611         212        239         122   
mean                  NaN         NaN        NaN         NaN   
std                   NaN         NaN        NaN         NaN   
min                   NaN         NaN        NaN         NaN   
25%                   NaN         NaN        NaN         NaN   
50%           

In [5]:
# Your code here
#austin.isnull().sum()
# great dataset

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [6]:
# Your code here
austin.DewPointAvgF.unique()

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

In [7]:
# Your observation here
# i see objects type, I do not know why pandas treat this like objects, but we can change it

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [8]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [9]:
# IMP!
# we'll run all columns performing an apply to check where are the non numeric values.
austin[wrong_type_columns]=austin[wrong_type_columns].apply(pd.to_numeric, errors="coerce")

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [10]:
# To check the type of each column
austin.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [11]:
# Your code here
# i do not see the usefulness of this:
# austin.isnull().any(axis=1)

# like i am not the robot cortocirciuto reading books. isnull().sum( ) is better
#austin.isnull().sum()

In [12]:
missing_values=austin.isnull().any(axis=1)
# selecting by col
austin["missing_values"]=missing_values

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [13]:
# I don't really understand it. missing values is a list as long as austin
len(missing_values)

1319

In [14]:
austin.shape

(1319, 22)

#### Calculate the ratio of missing rows to total rows

In [15]:
#st way:
missing_values.value_counts()
# we have 136 Trues and 1183 Falses. so:
ratio=136/1183
# that is equal to
ratio
# plz, uncertainties Vargas. UNCERTAINTIES.

0.11496196111580727

In [17]:
# Other way:
trues=0
for e in missing_values:
    if e:
        trues+=1
ratio=trues/len(missing_values)

print("the lengh is: ",(len(missing_values)),
      "\n the true values are: ",
      trues, "\n so the ratio is :",
      ratio,
     "\n Borgas plz, uncertainties, the real result is: ",
     "\n ratio= 0.1___ +- stuff to calculate")

the lengh is:  1319 
 the true values are:  136 
 so the ratio is : 0.10310841546626232 
 Borgas plz, uncertainties, the real result is:  
 ratio= 0.1___ +- stuff to calculate


As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [18]:
# Your code here
austin.columns

Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'PrecipitationSumInches', 'Events', 'missing_values'],
      dtype='object')

In [19]:
austin.Date.isna().sum()
austin.TempAvgF.isna().sum()
# to check

0

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [22]:
# Your code here
austin.PrecipitationSumInches.isna().sum()/len(missing_values)

# really? we do really want to remove this column because there are 10% of missing values? i think it 
# is too strict at first

0.09401061410159212

Over 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. It doesn't make sense to *guess* its missing values because the estimation will be too 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [23]:
# Your code here 
austin = austin.drop(["PrecipitationSumInches"], axis=1)

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [24]:
# Your code here
austin_fixed = austin.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [25]:
# Your code here
austin_fixed.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
missing_values                0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [26]:
# Your code here:
austin["Events"].value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [ ]:
# if the question is the max of different occurrences had happened in the same day:
# Fog, Rain, Thunderstom 

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [27]:
## IMP. ow to dummy columns in my dataframe
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
for i in event_list:
    austin_fixed[i] = 0

# Print your new dataframe to check whether new columns have been created:

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0                   57.0  ...                 2.0         20.0         4.0   
1                   43.0  ...                 5.0         16.0         6.0   
2                   27.0  ...                10.0          8.0         3.0   
3                   22.0  ...                 7.0         12.0         4.0   
4                   56.0  ...                 7.0         10.0         2.0   
...                  ...  ...                 ...          ...         ...   
1314                25.0  ...                10.0         12.0         5.0   
1315                20.0  ...                10.0         14.0         5.0   
1316                19.0  ...                10.0         12.0         4.0   
1317                27.0  ...                10.0         13.0         4.0   
1318                22.0  ...                10.0         12.0         4.0   

      WindGustMPH               Events  missing_values  Snow  Fog  Rain  \
0            31.0  Rain , Thunderstorm           False     0    0     0   
1            25.0                                False     0    0     0   
2            12.0                                False     0    0     0   
3            20.0                                False     0    0     0   
4            16.0                                 True     0    0     0   
...           ...                  ...             ...   ...  ...   ...   
1314         21.0                                False     0    0     0   
1315         20.0                                False     0    0     0   
1316         17.0                                False     0    0     0   
1317         20.0                                False     0    0     0   
1318         20.0                                False     0    0     0   

     Thunderstorm  
0               0  
1               0  
2               0  
3               0  
4               0  
...           ...  
1314            0  
1315            0  
1316            0  
1317            0  
1318            0  

[1319 rows x 25 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [29]:
# I know I could have recycle this. i'm just learning other thing
austin_fixed['Rain']=austin_fixed['Events'].str.contains('Rain').astype(int)
austin_fixed['Snow']=austin_fixed['Events'].str.contains('Snow').astype(int)
austin_fixed['Fog']=austin_fixed['Events'].str.contains('Fog').astype(int)
austin_fixed['Thunderstorm']=austin_fixed['Events'].str.contains('Thunderstorm').astype(int)

#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [30]:
# Your code here
austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
1     2013-12-22         56        48        39           43.0          36.0   
2     2013-12-23         58        45        32           31.0          27.0   
3     2013-12-24         61        46        31           36.0          28.0   
4     2013-12-25         58        50        41           44.0          40.0   
...          ...        ...       ...       ...            ...           ...   
1314  2017-07-27        103        89        75           71.0          67.0   
1315  2017-07-28        105        91        76           71.0          64.0   
1316  2017-07-29        107        92        77           72.0          64.0   
1317  2017-07-30        106        93        79           70.0          68.0   
1318  2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  ...  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0                   57.0  ...                 2.0         20.0         4.0   
1                   43.0  ...                 5.0         16.0         6.0   
2                   27.0  ...                10.0          8.0         3.0   
3                   22.0  ...                 7.0         12.0         4.0   
4                   56.0  ...                 7.0         10.0         2.0   
...                  ...  ...                 ...          ...         ...   
1314                25.0  ...                10.0         12.0         5.0   
1315                20.0  ...                10.0         14.0         5.0   
1316                19.0  ...                10.0         12.0         4.0   
1317                27.0  ...                10.0         13.0         4.0   
1318                22.0  ...                10.0         12.0         4.0   

      WindGustMPH               Events  missing_values  Snow  Fog  Rain  \
0            31.0  Rain , Thunderstorm           False     0    0     1   
1            25.0                                False     0    0     0   
2            12.0                                False     0    0     0   
3            20.0                                False     0    0     0   
4            16.0                                 True     0    0     0   
...           ...                  ...             ...   ...  ...   ...   
1314         21.0                                False     0    0     0   
1315         20.0                                False     0    0     0   
1316         17.0                                False     0    0     0   
1317         20.0                                False     0    0     0   
1318         20.0                                False     0    0     0   

     Thunderstorm  
0               1  
1               0  
2               0  
3               0  
4               0  
...           ...  
1314            0  
1315            0  
1316            0  
1317            0  
1318            0  

[1319 rows x 25 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [41]:
# Your code here
#perform this?

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [42]:
### DATETIME

from datetime import datetime as dt

austin_fixed.Date=austin_fixed.Date.map(lambda x: dt.strptime(x,'%Y-%m-%d').date().toordinal())

#### Print `austin_fixed` to check your `Date` column.

In [43]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0  735223         74        60        45           67.0          49.0   
1  735224         56        48        39           43.0          36.0   
2  735225         58        45        32           31.0          27.0   
3  735226         61        46        31           36.0          28.0   
4  735227         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0  ...                 7.0                 2.0         20.0         4.0   
1  ...                10.0                 5.0         16.0         6.0   
2  ...                10.0                10.0          8.0         3.0   
3  ...                10.0                 7.0         12.0         4.0   
4  ...                10.0                 7.0         10.0         2.0   

   WindGustMPH  missing_values  Snow  Fog  Rain  Thunderstorm  
0         31.0           False     0    0     1             1  
1         25.0           False     0    0     0             0  
2         12.0           False     0    0     0             0  
3         20.0           False     0    0     0             0  
4         16.0            True     0    0     0             0  

[5 rows x 24 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [44]:
# Your code here:
y=austin_fixed['TempAvgF']
X=austin_fixed[['TempHighF','TempLowF','DewPointHighF','DewPointAvgF','DewPointLowF','HumidityHighPercent','HumidityAvgPercent',
 'HumidityLowPercent','SeaLevelPressureHighInches','SeaLevelPressureAvgInches','SeaLevelPressureLowInches','VisibilityHighMiles','VisibilityAvgMiles',
 'VisibilityLowMiles','WindHighMPH','WindAvgMPH','WindGustMPH','Snow','Fog','Rain','Thunderstorm']]



In the next cell, import `train_test_split` from `sklearn.model_selection`

In [45]:
#Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [46]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [53]:
# Your code here:
austin_fixed.shape

(1319, 24)

In [52]:
ts_rows=1319*0.8
ts_rows

1055.2

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [51]:
# Your code here:
X_ts_train=X[:1055]
X_ts_test=X[1055:]

Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [50]:
# Your code here:
y_ts_train=y[:1055]
y_ts_test=y[1055:]


# Bonus Challenge 2

As explained in the README, the main purpose of this lab is to show you the typical process of preparing data for machine learning which sometimes takes up 90% of your time. Data cleaning is a valuable skill to learn and you need to be proficient at various techniques including the ones we showed you above as well as others you'll learn in the future. In the real world this skill will help you perform your job successfully and efficiently.

Now that we're done with data praparation, if you want to expeirence what you'll do in the rest 10% of your time, let's make the final leap.

We will use scikit-learn's [*Support Vector Machines*](https://scikit-learn.org/stable/modules/svm.html) to compute the fit of our training data set, the test on our test data set.

#### In the cell below, import `svm` from `sklearn`:

In [54]:
# Your code here
from sklearn import svm

#### Now, call `svm.SVC.fit()` on `X_train` and `y_train`. Assign the returned value to a variable called `clf` which stands for *classifier*. Then obtain the test score for `X_test` and `y_test` by calling `clf.score()`.

In [57]:
# Your code here
clf=svm.SVC()
clf.fit(X_train, y_train)
clf.score(X_test,y_test)
# no es un buen modelo

/home/vargas/miniconda3/envs/ironhack/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


0.0946969696969697

#### You now see the model fit score of your test data set. If it's extremely low, it means your selected model is not a good fit and you should try other models.

#### In addition to fitting `X_train`, `y_train`, `X_test`, and `y_test`, you can also fit `X_ts_train`, `y_ts_train`, `X_ts_test`, and `y_ts_test` if you completed Bonus Challenge 1.

In [56]:
# Your code here
clf=svm.SVC()
clf.fit(X_ts_train,y_ts_train)
clf.score(X_ts_test,y_ts_test)
#con esto se supone que solamente acierta el 9% de las veces, con lo cual sigue sin ser un buen modelo



/home/vargas/miniconda3/envs/ironhack/lib/python3.7/site-packages/sklearn/svm/base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


0.09090909090909091

#### We hope you have learned a lot of useful stuff in this lab!